src/apps/middleware.js

```diff
@@ -0,0 +1,54 @@
+import { UPDATE_LOCATION } from 'react-router-redux';
+
+import { newBundle, fetchBundleError } from './actions';
+import * as api from '../client-lib/api';
+import RendezvousSocket from '../client-lib/rendezvous-socket';
+
+// this should come from a config options, do this for now
+let outriggerHost = window.location.host;
+
+export const bundleMiddleware = store => {
+    let rendezvous;
+
+    function bundleReceived(bundle) {
+        api.describe(bundle)
+        .then(inputs => {
+            store.dispatch(newBundle(bundle, inputs));
+        })
+    }
+
+    function runModeChanged(runMode) {
+        if (runMode.path) {
+            api.getBundle(runMode.path)
+            .then(res => { bundleReceived(res.bundle) })
+            .catch(err => { store.dispatch(fetchBundleError(err)) })
+        } else if (runMode.rendezvous) {
+            if (rendezvous) {
+                rendezvous.close();
+            }
+
+            rendezvous = new RendezvousSocket(`ws://${outriggerHost}/rendezvous/${runMode.rendezvous}`);
+            rendezvous.on('message', msg => bundleReceived(msg.bundle));
+        }
+    }
+
+    runModeChanged(store.getState().runMode);
+
+    return next => action => {
+        // only update if prevRunMode is different from newRunMode
+        if (action.type !== UPDATE_LOCATION) {
+            return next(action);
+        }
+
+        let prevRunMode = store.getState().runMode;
+        let result =  next(action);
+        let newRunMode = store.getState().runMode;
```

Comment:
Darn you dave and your good questions. Fundamentally we're trying to initiate a set of behavior when a variable changes. I did a lot of thinking/research about this. I think ideally you would make the store some kind of observable like described here: https://github.com/rackt/redux/issues/303#issuecomment-125184409

I've seen some (very loose) precedence for doing what i'm doing in the logger middleware shown here: http://rackt.org/redux/docs/advanced/Middleware.html

I think I want to go ahead and continue with what I have because it works, and truthfully I don't fully understand the redux-observable stuff and don't really have time to think about it further.